### PR TITLE
Add TypeMarshaler interface

### PR DESCRIPTION
**Pull request by arv**
This allows go structs to participate in the MarshalType flow even
when they implement the Marshaler interface or if they have an original
tag.

### CODE REVIEW

File: go/marshal/encode_type.go [hunk 1]
```diff
@@ -54,13 +55,36 @@ func MustMarshalType(v interface{}) *types.Type {
...
+		if typ == nil {
+			panic(fmt.Errorf("nil result from %s.MarshalNomsType", t.String()))
+		}
+		return typ
+
+	}
+
 	if t.Implements(marshalerInterface) {
 		// There is no way to determine the noms type now. For Marshal it can be
 		// different each time MarshalNoms is called. This is handled further up the
```
Comment (kalman) [inline go/marshal/encode_type.go]: Could we actually generate an error here which says along the lines of `"Cannot marshal type which implements marshal.Marshaler, perhaps implement marshal.TypeMarshaler for $typename"`?

File: go/marshal/encode_type.go [hunk 2]
```diff
@@ -54,13 +55,36 @@ func MustMarshalType(v interface{}) *types.Type {
...
+		v := reflect.New(t)
+		typ, err := v.Interface().(TypeMarshaler).MarshalNomsType()
+		if err != nil {
+			panic(&marshalNomsError{err})
+		}
+		if typ == nil {
+			panic(fmt.Errorf("nil result from %s.MarshalNomsType", t.String()))
+		}
+		return typ
+
```
Comment (kalman) [inline go/marshal/encode_type.go]: extra blank line

File: go/marshal/encode_type.go (hunk 1 above)
Comment (arv) [inline go/marshal/encode_type.go]: Not at the moment. This is used by other parts of the system and they depend on nil. I could refactor to have those parts check for an error instead

Comment (arv) [inline go/marshal/encode_type.go]: I'm doing this... turns out to be simple.

File: go/marshal/encode_type.go [hunk 3]
```diff
@@ -68,24 +69,28 @@ var typeOfTypesType = reflect.TypeOf((*types.Type)(nil))
...
-
 	}
 
 	if t.Implements(marshalerInterface) {
+		if options.ReportErrors {
+			err := fmt.Errorf("Cannot marshal type which implements %s, perhaps implement %s for %s", marshalerInterface, typeMarshalerInterface, t)
+			panic(&marshalNomsError{err})
+		}
+
 		// There is no way to determine the noms type now. For Marshal it can be
```
Comment (kalman) [inline go/marshal/encode_type.go]: Move comment up

File: go/marshal/encode_type.go [hunk 4]
```diff
@@ -108,6 +113,12 @@ func encodeType(t reflect.Type, parentStructTypes []reflect.Type, tags nomsTags,
 		case "String":
 			return types.StringType
 		}
+
+		if options.ReportErrors {
+			err := fmt.Errorf("Cannot marshal type %s, it requires type parameters", t)
```
Comment (kalman) [inline go/marshal/encode_type.go]: I don't know what this error means

**Pull request merged**